app/services/fact_cleaner.rb

```diff
@@ -32,17 +32,19 @@ def delete_facts_names_values(fact_name_ids)
 
   def delete_orphaned_facts
     logger.debug "Cleaning leaf orphaned facts"
-    leaf_ids = []
-    FactName.leaves.includes(:fact_values).reorder(nil).find_in_batches(:batch_size => @batch_size) do |batch|
+    FactName.leaves.preload(:fact_values).reorder(nil).find_in_batches(:batch_size => @batch_size) do |batch|
       fact_name_ids = batch.select { |fact| fact.fact_values.empty? }
-      @deleted_count += delete_facts_names_values(fact_name_ids)
-      leaf_ids += fact_name_ids
+      @deleted_count += log_removed(FactName.unscoped.where(:id => fact_name_ids).delete_all, "fact name")
     end
 
     logger.debug "Cleaning compose orphaned facts"
-    FactName.composes.reorder(nil).find_in_batches(:batch_size => @batch_size) do |batch|
-      fact_name_ids = batch.select { |compose| compose.descendants.all? { |fact| leaf_ids.include?(fact.id) } }.map(&:id)
-      @deleted_count += delete_facts_names_values(fact_name_ids)
+    live_ancestors = Set.new
+    FactName.leaves.reorder(nil).select(:ancestry, :id).find_each(:batch_size => @batch_size) do |leaf|
+      live_ancestors.merge(leaf.ancestor_ids)
+    end
+    FactName.composes.reorder(nil).where.not(id: live_ancestors).select(:id).find_in_batches(:batch_size => @batch_size) do |batch|
```

Comment:
I like `select(:id)` can you also add it to the other queries?

Reply:
added to the pattern one, not sure if it will work well with the preload for the leafs one